### ISSUE TITLE
Release SecureDrop Workstation 0.3.1

### ISSUE THREAD

**Issue by eloquence** [freedomofpress/securedrop-workstation#571]
This is an issue to track the production release of the SecureDrop Workstation 0.3.1 RPM package, which ships the keyring update in #563. 

## RC preparation
- [x] Prepare 0.3.1-rc1 version - https://github.com/freedomofpress/securedrop-workstation/pull/570
- [x] Upload 0.3.1-rc1 packages to yum-test - https://github.com/freedomofpress/securedrop-workstation-dev-rpm-packages-lfs/pull/13

## Test plan: 
https://github.com/freedomofpress/securedrop-workstation/wiki/0.3.1-Test-Plan

## Release workflow
- [x] dom0: SecureDrop Workstation repo
  - [x] Bump version to 0.3.1 (final)
  - [x] Open PR to https://github.com/freedomofpress/securedrop-workstation/issues/571 with version bump
  - [x] Create git tag for 0.3.1
  - [x] Sign 0.3.1 tag with prod key
  - [x] Run `make dom0-rpm` from verified tag to create RPM - see full instructions [here](https://github.com/freedomofpress/securedrop-workstation#rpm-package)
  - [x] Submit prod RPM PR to https://github.com/freedomofpress/securedrop-workstation-prod-rpm-packages-lfs/
  - [x] Submit prod RPM build logs to https://github.com/freedomofpress/build-logs
  - [x] Publish packages to prod repo (via merge)
- [x] AppVMs: SecureDrop Debian Packaging repo
  - [x] Create git tag 0.2.10 (see instructions [here](https://github.com/freedomofpress/securedrop-workstation#tag-code-used-to-generate-artifacts))
  - [x] Sign 0.2.10 git tag with prod key
  - [x] Build packages from verified tag to create deb packages (see instructions [here](https://github.com/freedomofpress/securedrop-workstation#debian-package)):
    - [x] `PKG_VERSION=0.1.4 make securedrop-keyring`
    - [x] `PKG_VERSION=0.1.3 make securedrop-workstation-config`
  - [x]  Submit prod debs PR to https://github.com/freedomofpress/securedrop-debian-packages-lfs
  - [x] Submit prod debs build logs to https://github.com/freedomofpress/build-logs
  - [x] Publish packages to prod repo (via merge)

**Comment by conorsch** [comment]
When reporting the results of ad-hoc tests against SecureDrop Workstation, please include the following information:

## Environment:
- Environment type: _one of: Qubes staging
- Client version: N/A, did not perform functional testing
- Workstation version: 0.3.1rc1
- Server version: N/A, did not perform functional testing

## Test purpose:
Confirming keyring updates in Debian-based VMs

## Test steps:
Following test plan above

## Test results:
Success! Used the one-liner in the test plan above, and confirmed that _all_ VMs listed had the new 2021 key expiry. 

- [x] Make sure signing pubkey has is updated in TemplateVMs, via `sudo apt-key finger securedrop-release-key`.
  - [x] There should be only one key, with fingerprint 
  - [x] Expiration date is 2021-06-30 (_not_ 2020)

## Comments:
Proceeding with prod scenario next....

**Comment by eloquence** [comment]
## Staging report
Tested on a 0.3.0-staging enviroment that I upgraded to 0.3.1-rc1; then I ran `securedrop-admin --apply` to apply the configuration.

- [x] Make sure signing pubkey has is updated in TemplateVMs, via `sudo apt-key finger securedrop-release-key`.
  - [x] There should be only one key, with fingerprint 
  - [x] Expiration date is 2021-06-30 (_not_ 2020)

Detailed output from the very helpful one-liner here:
https://gist.github.com/eloquence/e9b4da12fb9bfeca2606d2def1b67bb5

```bash
$ grep 2020 output.txt | wc -l
0
$ grep 2021 output.txt | wc -l
14
```

**Comment by conorsch** [comment]
## Environment:
- Environment type: _one of: Qubes prod 
- Client version: N/A, did not perform functional testing
- Workstation version: 0.3.1rc1
- Server version: N/A, did not perform functional testing

## Test purpose:
Confirming keyring updates for RPM-based components, i.e. dom0 & sys-firewall

## Test steps:
Following test plan above for prod testing

## Test results:
:x: Test plan as described can't pass, needs to be rethought. The prod install logic references, naturally, the `apt.freedom.press` for TemplateVMs, but the `securedrop-keyring` package doesn't exist on `apt.freedom.press` yet, so the first-time install fails. 

<details>
<summary>Click to expand</summary>

```
  ----------
            ID: sd-log-remove-rsyslog-qubes-plugin
      Function: cmd.run
          Name: /rw/config/rc.local
        Result: False
       Comment: Command "/rw/config/rc.local" run
       Started: 17:25:02.434658
      Duration: 24.718 ms
       Changes:   
                ----------
                pid:
                    959
                retcode:
                    5
                stderr:
                    Failed to start securedrop-log.service: Unit securedrop-log.service not found.
                stdout:
  
  Summary for sd-log
  ------------
  Succeeded: 1 (changed=2)
  Failed:    1
  ------------
  Total states run:     2
  Total run time:  52.619 ms
Traceback (most recent call last):
  File "/usr/bin/securedrop-admin", line 67, in provision_all
    subprocess.check_call([os.path.join(SCRIPTS_PATH, "scripts/provision-all")])
  File "/usr/lib64/python3.5/subprocess.py", line 271, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/share/securedrop-workstation-dom0-config/scripts/provision-all']' returned non-zero exit status 20

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/securedrop-admin", line 171, in <module>
    main()
  File "/usr/bin/securedrop-admin", line 148, in main
    provision_all()
  File "/usr/bin/securedrop-admin", line 69, in provision_all
    raise SDAdminException("Error during provision-all")
__main__.SDAdminException: Error during provision-all
[user@dom0 ~]$ 


```

</details>


## Comments:

Typically for rc testing we'd use the procedure described in https://github.com/freedomofpress/securedrop-workstation/wiki/Workstation-dom0-QA-testing#upgrade-install-testing, but that approach would clobber the test key, rendering local testing moot.

**Comment by eloquence** [comment]
## Prod report

- Environment type: Qubes prod
- Client version: N/A, did not perform functional testing
- Workstation version: 0.3.0 fresh install->0.3.1rc1 from yum-test
- Server version: N/A, did not perform functional testing

tl;dr: While the `--apply` run threw errors as a result of the missing package, the critical portions of the test plan succeeded.

- [x] Performed fresh prod install
  - [x] `sudo dnf info securedrop-workstation-dom0-config` shows `0.3.0` in the version info
  - [x] `grep environment /usr/share/securedrop-workstation-dom0-config/config.json` shows `prod`
- [x] Manually install the rc1 package, following the _staging_ RPM instructions in https://github.com/freedomofpress/securedrop-workstation#staging-environment, then confirm you've got the proper version for testing:
  - [x] Run `sudo dnf info securedrop-workstation-dom0-config`
  - [x] Confirm `0.3.1` in the `Version` field
  - [x] Confirm `0.rc1.1.fc25` in the `Release` field
- [ ] Run `securedrop-admin --apply` to enforce the upgrade  :x: Run failed, see run output [here](https://gist.github.com/eloquence/4591b99170ad5d4c13a26f89a1ad57f6)
- [x] Make sure signing pukey is updated in dom0
   - [x] Check dom0 keyring:
     - [x] `mkdir -m 700 /tmp/fake-gpg-rpm-home`
     - [x] `gpg --homedir /tmp/fake-gpg-rpm-home --import /etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation`
     - [x] `gpg --homedir /tmp/fake-gpg-rpm-home -k`
     - [x] Confirm pubkey shows 2021-06-30 expiry
   - [x] Check sys-firewall keyring
       - [x] `qvm-run sys-firewall gnome-terminal` to open a terminal, then:
       - [x] `mkdir -m 700 /tmp/fake-gpg-rpm-home`
       - [x] `gpg --homedir /tmp/fake-gpg-rpm-home --import /etc/pki/rpm-gpg/RPM-GPG-KEY-securedrop-workstation`
       - [x] `gpg --homedir /tmp/fake-gpg-rpm-home -k`
       - [x] Confirm pubkey shows 2021-06-30 expiry

**Comment by conorsch** [comment]
Sounds good, thanks for the clarity, and for pressing ahead with the pubkey verification steps despite the failure in the `--apply` action, @eloquence. I'm able to confirm that the prod pubkey was added, despite the mismatch on the apt repository config leading to a failed task. 

Make a timeboxed attempt to revise the test plan for prod to use apt-test for domUs, but prod repos for dom0, via the following patch:

```
--- /srv/salt/sd-default-config.yml    2020-06-10 15:35:57.000000000 -0700
+++ /tmp/sd-default-config.yml    2020-06-15 09:42:28.600644786 -0700
@@ -2,8 +2,10 @@
 # Production variables, for use with real-world installs
 prod:
   dom0_yum_repo_url: "https://yum.securedrop.org/workstation/dom0/f25"
-  apt_repo_url: "https://apt.freedom.press"
-  signing_key_filename: "securedrop-release-signing-pubkey.asc"
+  #apt_repo_url: "https://apt.freedom.press"
+  #signing_key_filename: "securedrop-release-signing-pubkey.asc"
+  apt_repo_url: "https://apt-test.freedom.press"
+  signing_key_filename: "apt-test-pubkey.asc"
 # Development variables, suited for use during local development
 dev:
   dom0_yum_repo_url: "https://yum-test.securedrop.org/workstation/dom0/f25"
```
Unfortunately that's also insufficient for testing here—while the `securedrop-admin --apply` action completes without error, the gpg verification commands in the test plan show the test key for dom0, due to the patch on the `signing_key_filename` specifically. 

Given how thoroughly we've tested the various environments, I'm satisfied with results from rc1, and suggest proceeding with 0.3.1 final release. The OP has been updated with specific steps documenting the release workflow for 0.3.1.

**Comment by eloquence** [comment]
- [x] Run securedrop-admin --apply to enforce the upgrade 

This part is succeeded in my prod environment after applying updates, now that the prod deb packages are up (still using the RC1 RPM).

**Comment by conorsch** [comment]
Release is live.
